### ISSUE TITLE
[Windows] fluent-bit can not find log files grater then 2Gbytes

### ISSUE THREAD

**Issue by farcop** [fluent/fluent-bit#2208]
## Bug Report

**Describe the bug**
Using input plugin tail fluent-bit can not find log files grater then 2Gbytes

**To Reproduce**
We have file sized 2147483648bytes = 2Gbytes
```
PS D:\td-agent-bit\logs> dir


    Directory: D:\td-agent-bit\logs


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----       28.05.2020     19:24     2147483648 events1.log
```

Starting fluent-bit
```
> .\fluent-bit.exe -c ..\conf\fluent-bit.conf
Fluent Bit v1.5.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/05/28 19:47:51] [ info] Configuration:
[2020/05/28 19:47:51] [ info]  flush time     | 5.000000 seconds
[2020/05/28 19:47:51] [ info]  grace          | 5 seconds
[2020/05/28 19:47:51] [ info]  daemon         | 0
[2020/05/28 19:47:51] [ info] ___________
[2020/05/28 19:47:51] [ info]  inputs:
[2020/05/28 19:47:51] [ info]      tail
[2020/05/28 19:47:51] [ info] ___________
[2020/05/28 19:47:51] [ info]  filters:
[2020/05/28 19:47:51] [ info] ___________
[2020/05/28 19:47:51] [ info]  outputs:
[2020/05/28 19:47:51] [ info]      es.0
[2020/05/28 19:47:51] [ info] ___________
[2020/05/28 19:47:51] [ info]  collectors:
[2020/05/28 19:47:51] [debug] [storage] [cio stream] new stream registered: tail.0
[2020/05/28 19:47:51] [ info] [storage] version=1.0.3, initializing...
[2020/05/28 19:47:51] [ info] [storage] in-memory
[2020/05/28 19:47:51] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/05/28 19:47:51] [ info] [engine] started (pid=7384)
[2020/05/28 19:47:51] [debug] [engine] coroutine stack size: 98302 bytes (96.0K)
[2020/05/28 19:47:51] [debug] [input:tail:tail.0] scanning path D:\td-agent-bit\logs\events1.log
[2020/05/28 19:47:51] [debug] [input:tail:tail.0] 0 files found for 'D:\td-agent-bit\logs\events1.log'
[2020/05/28 19:47:51] [trace] [input:tail:tail.0 at C:\projects\fluent-bit-2e87g\plugins\in_tail\tail.c:217] scan path: D:\td-agent-bit\logs\events1.log
[2020/05/28 19:47:51] [debug] [output:es:es.0] host=101.1.1.132 port=9200 uri=/_bulk index=firid-test-1 type=sid
[2020/05/28 19:47:51] [debug] [router] match rule tail.0:es.0
[2020/05/28 19:47:51] [ info] [sp] stream processor started
```
There is no files found.

If we reduce this file to 2147483647 bytes size output will be:
```
...
[2020/05/28 19:54:12] [debug] [input:tail:tail.0] 1 files found for 'D:\td-agent-bit\logs\events1.log'
[2020/05/28 19:54:12] [trace] [input:tail:tail.0 at C:\projects\fluent-bit-2e87g\plugins\in_tail\tail.c:217] scan path: D:\td-agent-bit\logs\events1.log
...
```

**Expected behavior**
fluent-bit find and parse log files grater then 2Gbytes

**Environment**
Windows Server 2016 Standart
Tested on Fluent-bit 1.4.1 thue 1.5.0
```
[INPUT]
    Name   tail
    Tag    firid
    Parser firid
    Path   D:\td-agent-bit\logs\events*.log
    DB     ..\db\firid.db
```

**Comment by bluebike** [comment]
Is  **off_t**  signed 32 bit in Win32 API ?
https://github.com/fluent/fluent-bit/blob/560c4e38ebcaf23278bd0e940c2f9b3973e36235/plugins/in_tail/tail_file_internal.h#L38

Should **size** be using type **size_t** ?

**Comment by farcop** [comment]
@bluebike if we will use size_t on 32-bit API it gives us 4Gbyte limit as usigned variable. 
So question is Why fluent-bit use Win32 API on 64-bit operating system (Windows Server 2016)?

**Comment by bluebike** [comment]
@farcop   (I'm not Win32/64 expert)  If I underderstand right **off_t** is 32-bit in Win32 api.

**Comment by bluebike** [comment]
Actually maybe fluent-bit should have unit tests that checks that defined types can contain needed bits/bytes of information (?)
```c
TEST_CHECK(sizeof(size_t) >= 8);
TEST_CHECK(sizeof(off_t) >= 8);
...
```

**Comment by farcop** [comment]
@bluebike 
It seems you was right. It should be used size_t
https://docs.microsoft.com/en-us/cpp/c-runtime-library/standard-types?view=vs-2017
**off_t**, _off_t long integer | Represents file-offset value.
**size_t** (unsigned __int64 or unsigned integer, depending on the target platform) | Result of sizeof operator.

**Comment by bluebike** [comment]
Probably safest would define  size and  offset  as **uint64_t**.
Because older 32-bit linux/unix (like RHEL4)  has 32-bit size_t 
AND
Use version of lseek that can handle big files  (_lseeki64 in Win,  or _llseek in old unix/linux)

**Comment by bluebike** [comment]
I checked few old 32bit (i386) linuxes (RHEL 4,5,...)  and 
**size_t** and **off_t** where both 32 bit all of them.

So I thing this really policy/limit question: 
**Should fluent-bit (in_tail etc)  support over 2G files in 32bit systems?** 

If so... then probably best is to have  common file-size / seek function(s) ..
like ```int flb_lseek(fd,  int64_t offset, int whence,  uint64_t *result_offset) ```,
which would use normal lseek(2) in 64-bit unix systems,  _llseek,  lseek64 or  _lseeki64 (Win) in 32-bit systems.    Needs also some lines CMake magic.

**Comment by fujimotos** [comment]
@farcop @bluebike Thanks for investigation.

https://github.com/fluent/fluent-bit/blob/master/plugins/in_tail/tail_scan_win32.c#L73

The exact line failing is L73; stat() fails to return the meta info
for files > 2GB, and it confuses Fluent Bit to think that they are
not regular files.

As bluebike pointed out, there are several obstacles besides that
which prevent Fluent Bit from handling a large file.

I'll have a more closer look at this issue anytime soon, so WFM.

**Comment by farcop** [comment]
@bluebike @fujimotos Thank you!
We are looking forward to a solution of this issue.

**Comment by igorpeshansky** [comment]
It sounds like fluent-bit links against the 32-bit Windows APIs, which can't handle files over 2GiB in size. Linking against the 64-bit APIs should solve the issue. Maybe the mingw32 vs mingw64 detection is broken?

**Comment by fujimotos** [comment]
>  Linking against the 64-bit APIs should solve the issue. 
>  Maybe the mingw32 vs mingw64 detection is broken?

@igorpeshansky Sadly it is not exactly like that. `stat()` on MSVC
uses 32-bit size even on 64bit platform.

https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/stat-functions?view=vs-2019

To handle large files, we need to explicitly use `stati64()`.
Such implicit limitations hide there and here.

**Comment by AlekseyKalinin** [comment]
We have same issue. Someone could say when fix will be implemented?

**Comment by fujimotos** [comment]
For what it's worth, I'm planning to work on this issue ("large file support"
in the next v1.6 development cycle.

**Comment by fujimotos** [comment]
> We have same issue. Someone could say when fix will be implemented?

@AlekseyKalinin I have done adding 64-bit file support to `in_tail` plugin.

This issue should be resolved once #2584 landed onto the main line.
(I expect that the fix will be included in the next stable version: v1.6.0)

**Comment by AlekseyKalinin** [comment]
> This issue should be resolved once #2584 landed onto the main line.
> (I expect that the fix will be included in the next stable version: v1.6.0)

Great news! We're looking forward this version.
